docs/CONTRIBUTING.md

```diff
@@ -12,6 +12,122 @@ Take a look at the [MongoDB connector](../connectors/sources/mongo.py) for more
+
+#### How to implement advanced rules
+
+When implementing a new connector you've to follow the API of the [BaseDataSource](../connectors/source.py).
+The custom implementation for advanced rules is usually located inside the `get_docs` function:
+
+```python
+async def get_docs(self, filtering=None):
+    if filtering and filtering.has_advanced_rules():
+        advanced_rules = filtering.get_advanced_rules()
+        # your custom advanced rules implementation
+    else:
+        # default fetch all data implementation 
+```
+
+Here you can pass f.e. custom queries to a database or something similar.
+The structure of the advanced rules is dependent on your implementation and your concrete use case.
+For MySQL the advanced rules structure looks like this:
+You specify databases on the top level, which contain tables, which specify a custom query.
+
+Example:
+
+```json
+{
+  "database_1": {
+    "table_1": "SELECT ... FROM ...;",
+    "table_2": "SELECT ... FROM ...;"
+  },
+  "database_2": {
+    "table_1": "SELECT ... FROM ...;"
+  }
+}
+```
+
+Note that the framework calls `get_docs` with the parameter `filtering` being of type `Filter`, which is located in [byoc.py](../connectors/sources/mysql.py).
+The `Filter` class provides convenient methods to extract advanced rules from the filter and to check whether advanced rules are present.
+
+#### How to validate advanced rules
+
+For validating advanced rules the framework takes the list of validators returned by the method `advanced_rules_validators` and calls them in the order they appear in that list.
+By default, this list is empty in the [BaseDataSource](../connectors/source.py) as advanced rules are always specific to the connector implementation.
+You can plug in custom validators by implementing a class containing a `validate` method, which accepts one parameter.
+The framework expects the custom validators to return a `SyncRuleValidationResult`, which can be found in [validation.py](../connectors/filtering/validation.py).
+
+```python
+class MyValidator(AdvancedRulesValidator):
+    
+    def validate(self, advanced_rules):
+        # custom validation logic
+        return SyncRuleValidationResult(...)
+```
+
+Note that the framework will call `validate` with the parameter `advanced_rules` being of type `dict`.
+Now you can return a list of validator instances in `advanced_rules_validators`:
+
+```python
+class MyDataSource(BaseDataSource):
+
+    def advanced_rules_validators(self):
+        return [MyValidator()]
+```
+
+The framework will handle the rest like scheduling validation, calling the custom validators and storing the corresponding results.
+
+#### How to provide custom basic rule validation
+
+The framework already provides default validations for basic rules.
+If you want to extend the default validation you can provide custom basic rules validators.
```

Comment:
```suggestion
To extend the default validation, provide custom basic rules validators.
```

Reply:
Again, cutting out these introductory phrases makes the sentence much easier to read.